File: GSEA.and.intersections/run_fgsea_DESeq2_databases_GSEA.R
# =============================================================================
# FGSEA using DESeq2 CSV results + GMT files in databases_GSEA
# =============================================================================
# - Reads DESeq2 CSVs (Gene_symbol, log2FoldChange, padj, ...) from project root
# - Ranks genes by log2FoldChange (or stat if you prefer)
# - Loads GMTs from ../databases_GSEA/ (Hallmark, KEGG, Reactome, WikiPathways)
# - Runs fgsea per cell type × pathway set; saves tables and plots
# =============================================================================

library(tidyverse)
library(fgsea)

# ----- Paths (run from scripts_FGSEA or project root) -----
base_dir <- if (dir.exists("databases_GSEA")) getwd() else dirname(getwd())
csv_dir <- base_dir
gmt_dir <- file.path(base_dir, "databases_GSEA")
out_dir <- file.path(base_dir, "GSEA_results_DESeq2")
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
dir.create(file.path(out_dir, "tables"), showWarnings = FALSE)
dir.create(file.path(out_dir, "plots"), showWarnings = FALSE)

# ----- DESeq2 CSV files (cell-type names from filenames) -----
csv_files <- list.files(csv_dir, pattern = "\\.csv$", full.names = TRUE)
csv_files <- csv_files[basename(csv_files) != "summary_DEG_counts.csv"]
if (length(csv_files) == 0) {
  stop("No CSV files found in ", csv_dir)
}
cat("DESeq2 CSV files to process:", length(csv_files), "\n")
cat("GMT directory:", gmt_dir, "\n\n")

# ----- GMT files in databases_GSEA -----
gmt_sets <- list(
  Hallmark     = file.path(gmt_dir, "h.all.v2026.1.Hs.symbols.gmt"),
  KEGG         = file.path(gmt_dir, "c2.cp.kegg_medicus.v2026.1.Hs.symbols.gmt"),
  Reactome     = file.path(gmt_dir, "c2.cp.reactome.v2026.1.Hs.symbols.gmt"),
  WikiPathways = file.path(gmt_dir, "c2.cp.wikipathways.v2026.1.Hs.symbols.gmt"),
  GO_BP        = file.path(gmt_dir, "c5.go.bp.v2026.1.Hs.symbols.gmt")
)
gmt_sets <- gmt_sets[vapply(gmt_sets, file.exists, logical(1))]
if (length(gmt_sets) == 0) {
  stop("No GMT files found in ", gmt_dir)
}
cat("GMT sets to use:", paste(names(gmt_sets), collapse = ", "), "\n\n")

# ----- Process each DESeq2 CSV -----
for (csv_path in csv_files) {
  cell_type <- tools::file_path_sans_ext(basename(csv_path))
  cat("======== ", cell_type, " ========\n", sep = "")

  res <- read_csv(csv_path, col_types = cols(), show_col_types = FALSE)
  if (!"Gene_symbol" %in% names(res)) {
    cat("  Skip: no Gene_symbol column.\n\n")
    next
  }
  if (!"log2FoldChange" %in% names(res)) {
    cat("  Skip: no log2FoldChange column.\n\n")
    next
  }

  # Rank by log2FoldChange (DESeq2 style); drop NA and duplicate symbols
  res2 <- res %>%
    dplyr::select(Gene_symbol, log2FoldChange) %>%
    na.omit() %>%
    distinct() %>%
    group_by(Gene_symbol) %>%
    summarise(log2FoldChange = mean(log2FoldChange), .groups = "drop") %>%
    arrange(desc(log2FoldChange))
  ranks <- deframe(res2)
  cat("  Ranked genes:", length(ranks), "\n")

  for (set_name in names(gmt_sets)) {
    gmt_file <- gmt_sets[[set_name]]
    pathways <- gmtPathways(gmt_file)
    cat("  Running fgsea:", set_name, "... ")
    fgseaRes <- fgsea(pathways = pathways, stats = ranks, nperm = 1000)
    fgseaResTidy <- fgseaRes %>% as_tibble() %>% arrange(desc(NES))

    out_table <- file.path(out_dir, "tables",
                           paste0(cell_type, "_", set_name, "_fgsea.csv"))
    write_csv(fgseaResTidy, out_table)

    # Barplot (top pathways by NES)
    n_plot <- min(30, nrow(fgseaResTidy))
    p <- ggplot(head(fgseaResTidy, n_plot), aes(reorder(pathway, NES), NES)) +
      geom_col(aes(fill = padj < 0.05)) +
      coord_flip() +
      labs(x = "", y = "NES", title = paste(cell_type, "-", set_name)) +
      theme_minimal() +
      theme(axis.text.y = element_text(size = 8))
    ggsave(file.path(out_dir, "plots",
                     paste0(cell_type, "_", set_name, "_barplot.png")),
           p, width = 10, height = max(6, n_plot * 0.2), dpi = 150, bg = "white")
    cat("OK\n")
  }
  cat("\n")
}

cat("Done. Results in:", out_dir, "\n")
cat("  tables/: *_fgsea.csv\n")
cat("  plots/:  *_barplot.png\n")

File: GSEA.and.intersections/run_fgsea_DESeq2_databases_GSEA.py
#!/usr/bin/env python3
# =============================================================================
# FGSEA (prerank GSEA) using DESeq2 CSV results + GMT files in databases_GSEA
# =============================================================================
# - Reads DESeq2 CSVs (Gene_symbol, log2FoldChange, padj, ...) from project root
# - Ranks genes by log2FoldChange
# - Loads GMTs from databases_GSEA (Hallmark, KEGG, Reactome, WikiPathways, GO_BP)
# - Runs gseapy.prerank per cell type x pathway set; saves tables and plots
# Requires: pandas, gseapy, matplotlib
# =============================================================================

import os
import sys
from pathlib import Path

import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

try:
    import gseapy as gp
except ImportError:
    print("ERROR: gseapy not installed. Install with: pip install gseapy")
    sys.exit(1)

# ----- Paths (run from scripts_FGSEA or project root) -----
if Path("databases_GSEA").is_dir():
    base_dir = Path.cwd()
else:
    base_dir = Path.cwd().parent
csv_dir = base_dir
gmt_dir = base_dir / "databases_GSEA"
out_dir = base_dir / "GSEA_results_DESeq2_python"
out_dir.mkdir(exist_ok=True)
(out_dir / "tables").mkdir(exist_ok=True)
(out_dir / "plots").mkdir(exist_ok=True)

# ----- DESeq2 CSV files -----
csv_files = sorted(csv_dir.glob("*.csv"))
csv_files = [f for f in csv_files if f.name != "summary_DEG_counts.csv"]
if not csv_files:
    print("ERROR: No CSV files found in", csv_dir)
    sys.exit(1)
print("DESeq2 CSV files to process:", len(csv_files))
print("GMT directory:", gmt_dir, "\n")

# ----- GMT files in databases_GSEA -----
gmt_sets = {
    "Hallmark":     gmt_dir / "h.all.v2026.1.Hs.symbols.gmt",
    "KEGG":         gmt_dir / "c2.cp.kegg_medicus.v2026.1.Hs.symbols.gmt",
    "Reactome":     gmt_dir / "c2.cp.reactome.v2026.1.Hs.symbols.gmt",
    "WikiPathways": gmt_dir / "c2.cp.wikipathways.v2026.1.Hs.symbols.gmt",
    "GO_BP":        gmt_dir / "c5.go.bp.v2026.1.Hs.symbols.gmt",
}
gmt_sets = {k: v for k, v in gmt_sets.items() if v.is_file()}
if not gmt_sets:
    print("ERROR: No GMT files found in", gmt_dir)
    sys.exit(1)
print("GMT sets to use:", ", ".join(gmt_sets.keys()), "\n")

# ----- Process each DESeq2 CSV -----
for csv_path in csv_files:
    cell_type = csv_path.stem
    print("======== ", cell_type, " ========", sep="")

    res = pd.read_csv(csv_path)
    if "Gene_symbol" not in res.columns:
        print("  Skip: no Gene_symbol column.\n")
        continue
    if "log2FoldChange" not in res.columns:
        print("  Skip: no log2FoldChange column.\n")
        continue

    # Rank by log2FoldChange; drop NA, dedupe by mean; uppercase symbols for GMT match
    res2 = res[["Gene_symbol", "log2FoldChange"]].dropna()
    res2["Gene_symbol"] = res2["Gene_symbol"].astype(str).str.upper()
    rnk = (
        res2.groupby("Gene_symbol", as_index=True)["log2FoldChange"]
        .mean()
        .sort_values(ascending=False)
    )
    print("  Ranked genes:", len(rnk))

    for set_name, gmt_file in gmt_sets.items():
        print("  Running prerank:", set_name, "... ", end="", flush=True)
        try:
            pr = gp.prerank(
                rnk=rnk,
                gene_sets=str(gmt_file),
                outdir=None,
                permutation_num=1000,
                no_plot=True,
                seed=42,
                min_size=10,
                max_size=500,
            )
            # gseapy .res2d: pathway names are in "Term". A column sometimes named "pathway"
            # can wrongly contain the method label "prerank" for every row — use Term for labels.
            df = pr.res2d.copy()
            if "Term" in df.columns:
                df["pathway"] = df["Term"].astype(str)
                df = df.drop(columns=["Term"], errors="ignore")
            elif "Name" in df.columns:
                df = df.rename(columns={"Name": "pathway"})
            # p-value columns (gseapy may use NOM p-val / FDR q-val)
            if "NOM p-val" in df.columns and "pval" not in df.columns:
                df = df.rename(columns={"NOM p-val": "pval"})
            if "FDR q-val" in df.columns and "padj" not in df.columns:
                df = df.rename(columns={"FDR q-val": "padj"})
            df = df.sort_values("NES", ascending=False)

            out_table = out_dir / "tables" / f"{cell_type}_{set_name}_fgsea.csv"
            df.to_csv(out_table, index=False)

            # Barplot (top pathways by NES)
            n_plot = min(30, len(df))
            plot_df = df.head(n_plot)
            padj_col = "padj" if "padj" in plot_df.columns else "pval"
            fig, ax = plt.subplots(figsize=(10, max(6, n_plot * 0.2)))
            y_pos = range(len(plot_df))
            nes = plot_df["NES"].values
            colors = ["#e41a1c" if p < 0.05 else "#377eb8" for p in plot_df[padj_col]]
            ax.barh(y_pos, nes, color=colors)
            ax.set_yticks(y_pos)
            ax.set_yticklabels(plot_df["pathway"].tolist(), fontsize=8)
            ax.set_xlabel("NES")
            ax.set_ylabel("")
            ax.set_title(f"{cell_type} - {set_name}")
            ax.invert_yaxis()
            plt.tight_layout()
            plt.savefig(
                out_dir / "plots" / f"{cell_type}_{set_name}_barplot.png",
                dpi=150,
                bbox_inches="tight",
            )
            plt.close()
            print("OK")
        except Exception as e:
            print("ERROR:", e)

    print()

print("Done. Results in:", out_dir)
print("  tables/: *_fgsea.csv")
print("  plots/:  *_barplot.png")
